From 519bf96186ff538b9230b9362baff08932b5a8f8 Mon Sep 17 00:00:00 2001
From: Maxim Yu. Osipov <mosipov@dev.rtsoft.ru>
Date: Fri, 19 Apr 2013 14:16:43 +0400
Subject: [PATCH 071/127] arm: tegra: ktt30: Added support of Power button.

This adds support of power button with the help of gpio_keys
infrastructure. Single push on the running board generates event
via /dev/input/<event> interface, which could be processed
by application layer accordingly (e.g. as a signal to shutdown the board).
When the board is in suspended mode, short push will wakeup the board.

Signed-off-by: Maxim Osipov <mosipov@dev.rtsoft.ru>
Signed-off-by: Valentine Barshak <vbarshak@dev.rtsoft.ru>
---
 arch/arm/mach-tegra/board-ktt30.c |   37 +++++++++++++++++++++++++++++++++++++
 1 files changed, 37 insertions(+), 0 deletions(-)

diff --git a/arch/arm/mach-tegra/board-ktt30.c b/arch/arm/mach-tegra/board-ktt30.c
index 199ab05..7d63899 100644
--- a/arch/arm/mach-tegra/board-ktt30.c
+++ b/arch/arm/mach-tegra/board-ktt30.c
@@ -31,6 +31,7 @@
 #include <linux/delay.h>
 #include <linux/i2c-tegra.h>
 #include <linux/gpio.h>
+#include <linux/gpio_keys.h>
 #include <linux/input.h>
 #include <linux/platform_data/tegra_usb.h>
 #include <linux/spi/spi.h>
@@ -71,6 +72,17 @@
 #include "pm.h"
 #include "wdt-recovery.h"
 
+#define GPIO_IKEY(_id, _irq, _iswake, _deb)	\
+	{					\
+		.code = _id,			\
+		.gpio = -1,			\
+		.irq = _irq,			\
+		.desc = #_id,			\
+		.type = EV_KEY,			\
+		.wakeup = _iswake,		\
+		.debounce_interval = _deb,	\
+	}
+
 static struct balanced_throttle throttle_list[] = {
 #ifdef CONFIG_TEGRA_THERMAL_THROTTLE
 	{
@@ -519,6 +531,30 @@ static struct platform_device ktt30_audio_wm8903_device = {
 	},
 };
 
+static struct gpio_keys_button ktt30_int_keys[] = {
+	[0] = GPIO_IKEY(KEY_POWER,
+		 TPS6591X_IRQ_BASE + TPS6591X_INT_PWRON, 1, 100)
+};
+
+static struct gpio_keys_platform_data ktt30_int_keys_pdata = {
+	.buttons	= ktt30_int_keys,
+	.nbuttons       = ARRAY_SIZE(ktt30_int_keys),
+};
+
+static struct platform_device ktt30_int_keys_device = {
+	.name   = "gpio-keys",
+	.id     = 0,
+	.dev    = {
+		.platform_data  = &ktt30_int_keys_pdata,
+	},
+};
+
+int __init ktt30_keys_init(void)
+{
+	platform_device_register(&ktt30_int_keys_device);
+	return 0;
+}
+
 static struct platform_device *ktt30_devices[] __initdata = {
 	&tegra_pmu_device,
 	&tegra_rtc_device,
@@ -817,6 +853,7 @@ static void __init tegra_ktt30_init(void)
 #ifdef CONFIG_TEGRA_WDT_RECOVERY
 	tegra_wdt_recovery_init();
 #endif
+	ktt30_keys_init();
 }
 
 static void __init tegra_ktt30_reserve(void)
-- 
1.7.4.4

